pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,882 @@
+	var resultLock sync.Mutex
+	var candidates []defaultpreemption.Candidate
+	snapshotState, err := getSnapshotState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", SnapshotStateKey, err)
+		return nil
+	}
+	preFilterState, err := getPreFilterState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", preFilterStateKey, err)
+		return nil
+	}
+
+	checkNode := func(i int) {
+		nodeInfoCopy := potentialNodes[i].Clone()
+		stateCopy := state.Clone()
+		pods, numPDBViolations, fits := selectVictimsOnNode(ctx, fh, stateCopy, pod, nodeInfoCopy, pdbs, snapshotState, preFilterState)
+		if fits {
+			resultLock.Lock()
+			victims := extenderv1.Victims{
+				Pods:             pods,
+				NumPDBViolations: int64(numPDBViolations),
+			}
+			c := candidate{
+				victims: &victims,
+				name:    nodeInfoCopy.Node().Name,
+			}
+			candidates = append(candidates, &c)
+			resultLock.Unlock()
+		}
+	}
+	cutil.Until(ctx, len(potentialNodes), checkNode)
+	return candidates
+}
+
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// that may be satisfied by removing pods from the node.
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+	var potentialNodes []*framework.NodeInfo
+	for _, node := range nodes {
+		name := node.Node().Name
+		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
+		// to determine whether preemption may help or not on the node.
+		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			continue
+		}
+		potentialNodes = append(potentialNodes, node)
+	}
+	return potentialNodes
+}
+
+func selectVictimsOnNode(
+	ctx context.Context,
+	ph framework.PreemptHandle,
+	state *framework.CycleState,
+	pod *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+	snapshotState *SnapshotState,
+	preFilterState *PreFilterState,
+) ([]*v1.Pod, int, bool) {
+	var potentialVictims []*v1.Pod
+	var moreThanMinWithPreemptor bool
+
+	removePod := func(rp *v1.Pod) error {
+		if err := nodeInfo.RemovePod(rp); err != nil {
+			return err
+		}
+		status := ph.RunPreFilterExtensionRemovePod(ctx, state, pod, rp, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+	addPod := func(ap *v1.Pod) error {
+		nodeInfo.AddPod(ap)
+		status := ph.RunPreFilterExtensionAddPod(ctx, state, pod, ap, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	podPriority := podutil.GetPodPriority(pod)
+	preemptorElasticQuotaInfo, preemptorWithElasticQuota := snapshotState.elasticQuotas[pod.Namespace]
+
+	// Check if there is elastic quota in the preemptor's namespace.
+	if preemptorWithElasticQuota {
+		moreThanMinWithPreemptor = preemptorElasticQuotaInfo.moreThanMinWithPod(preFilterState.Resource)
+	}
+
+	for _, p := range nodeInfo.Pods {
```

Comment:
I actually expected a more efficient algorithm to loop over pods by elastic quota's namespace.